onnxruntime/core/providers/dml/dml_provider_factory.cc

```diff
@@ -216,6 +265,128 @@
   API_IMPL_END
 }
 
+ORT_API_STATUS_IMPL(FreeGPUAllocation_2, _In_ void* ptr) {
+  API_IMPL_BEGIN
+#ifdef USE_DML
+  Dml::FreeGPUAllocation(ptr);
+#endif  // USE_DML
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtSessionOptionsAppendExecutionProvider_DML2, _In_ OrtSessionOptions* options, OrtDmlDeviceOptions* device_opts) {
+API_IMPL_BEGIN
+  //options->provider_factories.push_back(onnxruntime::DMLProviderFactoryCreator::Create(device_id));
+  
+    OrtDmlPerformancePreference p = device_opts->p;
+    OrtDmlDeviceFilter f = device_opts->f;
+
+    // Create DXCore Adapter Factory
+    ComPtr<IDXCoreAdapterFactory> adapterFactory;
+    ORT_THROW_IF_FAILED(::DXCoreCreateAdapterFactory(adapterFactory.GetAddressOf()));
+
+    // Get a list of all the adapters that support compute
+    ComPtr<IDXCoreAdapterList> d3D12CoreComputeAdapters;
+    GUID attributes[]{ DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE };
+    ORT_THROW_IF_FAILED(
+        adapterFactory->CreateAdapterList(_countof(attributes),
+            attributes,
+            d3D12CoreComputeAdapters.GetAddressOf()));
+
+    const uint32_t count{ d3D12CoreComputeAdapters->GetAdapterCount() };
+    int compute_only_device_id = -1;
+
+    // will hold all the adapters in the order specified by the device options
+    ComPtr<IDXCoreAdapter>* ordered_adapters =
+        (ComPtr<IDXCoreAdapter>*) malloc(count * sizeof(ComPtr<IDXCoreAdapter>));
+    int ordered_adapter_count = 0;
+
+    // Used in the case that both GPU and NPU adapters are considered
+    ComPtr<IDXCoreAdapter>* gpu_adapters =
+        (ComPtr<IDXCoreAdapter>*) malloc(count * sizeof(ComPtr<IDXCoreAdapter>));
+    int gpu_adapter_count = 0;
+    ComPtr<IDXCoreAdapter>* npu_adapters =
+        (ComPtr<IDXCoreAdapter>*) malloc(count * sizeof(ComPtr<IDXCoreAdapter>));
```

Comment:
rather than having three lists,
it may be better to create a temporary struct:
struct AdapterInfo {
  ComPtr<IDXCoreAdapter> Adapter;
  OrtDmlDeviceFilter Filter; // GPU or NPU
}